### PR TITLE
parser: allow `match` nodes w/o `when` if there is an `else`

### PR DESCRIPTION
**Pull request by Kijewski**

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Nice idea! Can you update the book to mention `else` for `match` please?

**Comment by Kijewski** [comment]
> Nice idea!

First I only wanted to give a better message if there are no `when` blocks, but then I couldn't think of a good reason why not to allow it if there is an `else`.

> Can you update the book to mention `else` for `match` please?

There is a sentence "The `else` block is equivalent to matching on `_` (matching anything)", but I will try if I can word it better.

File: book/src/template_syntax.md [hunk 1]
```diff
@@ -559,16 +559,49 @@ expand an `Option`:
...
+{% endmatch %}
+```
+
+Or some more complex type, such as a [`Result<T, E>`](https://doc.rust-lang.org/stable/std/result/enum.Result.html):
+
+```jinja
+{% match result %}
+  {% when Ok(val) %} Good: {{ val }}.
+  {% when Err(err) %} Bad: {{ err }}.
+```
```
Comment (GuillaumeGomez) [inline book/src/template_syntax.md:582]:
``````suggestion
{% endmatch %}
```
``````

File: book/src/template_syntax.md [hunk 2]
```diff
@@ -559,16 +559,49 @@ expand an `Option`:
...
+
+Using the placeholder `_` to match against any value without capturing the datum, works too.
+The wildcard operator `..` is used to match against an arbitrary amount of items,
+and the same restrictions as in Rust, e.g. that it can be used only once in a slice or struct:
+
+```jinja
+{% match list_of_ints %}
+  {% when [first, ..] %} The list starts with a {{ first }}
+  {% else %} The list is empty.
```
Comment (GuillaumeGomez) [inline book/src/template_syntax.md:591]: I would use `{% when _ %}` here and then rewrite this example after the `{% else %}` explanation below.

Comment (Kijewski) [reply book/src/template_syntax.md]: Changed.

**Comment by Kijewski** [comment]
Huh, the alternatives syntax `{% when 1 | 2 | 3 %}` does not actually work, but I am sure I saw it in the code.

**Review by GuillaumeGomez — APPROVED** [review]

File: book/src/template_syntax.md [hunk 3]
```diff
@@ -559,16 +559,55 @@ expand an `Option`:
...
+A `{% match %}` must be exhaustive, i.e. all possible inputs must have a case.
+This is most easily done by using proving an `{% else %}` case,
+if not all possible values need an individual handling.
+
+```jinja
+{% match answer %}
+  {% when Ok(42) %} The answer is "42".
+  {% else %} No answer wrong answer?
+{% endmatch %}
+```
```
Comment (GuillaumeGomez) [inline book/src/template_syntax.md:607]:
``````suggestion
The `{% else %}` node is syntactical sugar for `{% when _ %}`.
If used, it must come last, after all other `{% when %}` blocks:

```jinja
{% match answer %}
  {% when Ok(42) %} The answer is "42".
  {% else %} No answer wrong answer?
{% endmatch %}
```

A `{% match %}` must be exhaustive, i.e. all possible inputs must have a case.
This is most easily done by using proving an `{% else %}` case,
if not all possible values need an individual handling.
``````

**Comment by GuillaumeGomez** [comment]
Once my last comment is done, please merge. :)

**Comment by GuillaumeGomez** [comment]
I opened https://github.com/rinja-rs/rinja/issues/130 for the `|` in `when`.

**Comment by Kijewski** [comment]
I hoped if I use the "apply suggestion" button it might not dismiss your review, but alas, it does. :)

**Comment by GuillaumeGomez** [comment]
Hehehe. I can click on the "approve" button again. :3

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks for the improvement!

**Pull request merged**